rbc/omnisci_pipeline.py

```diff
@@ -0,0 +1,107 @@
+from numba.core.untyped_passes import IRProcessing
+from numba import types
+from llvmlite import ir as llvm_ir
+from .omnisci_array import builder_buffers, ArrayPointer
+
+int8_t = llvm_ir.IntType(8)
+int32_t = llvm_ir.IntType(32)
+int64_t = llvm_ir.IntType(64)
+void_t = llvm_ir.VoidType()
+
+
+@extending.intrinsic
+def free_omnisci_array(typingctx, ret):
+    sig = types.void(ret)
+
+    def codegen(context, builder, signature, args):
+        buffers = builder_buffers[builder]
+
+        # TODO: using stdlib `free` that works only for CPU. For CUDA
+        # devices, we need to use omniscidb provided deallocator.
+        free_fnty = llvm_ir.FunctionType(void_t, [int8_t.as_pointer()])
+        free_fn = builder.module.get_or_insert_function(free_fnty, name="free")
+
+        # We skip the ret pointer iff we're returning an Array
+        # otherwise, we free everything
+        if isinstance(ret, ArrayPointer):
+            [arg] = args
+            ptr = builder.load(builder.gep(arg, [int32_t(0), int32_t(0)]))
+            ptr = builder.bitcast(ptr, int8_t.as_pointer())
+            for ptr8 in buffers:
+                with builder.if_then(builder.icmp_signed('!=', ptr, ptr8)):
+                    cgutils.printf(builder, "[Free] Deleting ptr: %p\n", ptr8)
+                    builder.call(free_fn, [ptr8])
+                with builder.if_then(builder.icmp_signed('==', ptr, ptr8)):
+                    cgutils.printf(builder, "[Free] Skipping ptr %p\n", ptr)
+        else:
+            for ptr8 in buffers:
+                cgutils.printf(builder, "[Free] Deleting ptr: %p\n", ptr8)
+                builder.call(free_fn, [ptr8])
+
+        del builder_buffers[builder]
+
+    return sig, codegen
+
+
+# Register this pass with the compiler framework, declare that it will not
+# mutate the control flow graph and that it is not an analysis_only pass (it
+# potentially mutates the IR).
+@register_pass(mutates_CFG=False, analysis_only=False)
+class FreeOmnisciArray(FunctionPass):
+    _name = "free_omnisci_arrays"  # the common name for the pass
+
+    def __init__(self):
+        FunctionPass.__init__(self)
+
+    # implement method to do the work, "state" is the internal compiler
+    # state from the CompilerBase instance.
+    def run_pass(self, state):
+        func_ir = state.func_ir  # get the FunctionIR object
+        found = False
+
+        for blk in func_ir.blocks.values():
+            for stmt in blk.find_insts(ir.Assign):
+                if isinstance(stmt.value, ir.FreeVar) \
+                   and stmt.value.name == 'Array':
+                    found = True
+
+        if not found:
+            return False  # we do not change the IR
+
+        for blk in func_ir.blocks.values():
+            loc = blk.loc
+            scope = blk.scope
+            for ret in blk.find_insts(ir.Return):
+
+                name = 'free_omnisci_array_fn'
+                value = ir.Global(name, free_omnisci_array, loc)
```

Comment:
nit: To eliminate `name`, use
```
value = ir.Global(free_omnisci_array.__name__, free_omnisci_array, loc)
```

Reply:
I just did that because flake8 complains otherwise, due to the characters line limit

Reply:
Btw, it seems that intrinsics do not have a `__name__` property.

Reply:
ok, nvm then.